src/platform/packages/shared/kbn-test/src/jest/run.ts

```diff
@@ -137,31 +148,79 @@ export function runJest(configName = 'jest.config.js') {
     }
 
     log.verbose(`no config provided, found ${configPath}`);
+    resolvedConfigPath = configPath;
     process.argv.push('--config', configPath);
 
     if (!testFilesProvided) {
       log.verbose(`no test files provided, setting to current directory`);
       process.argv.push(relative(wd, cwd));
     }
-
-    log.info('yarn jest', process.argv.slice(2).join(' '));
-
-    if (SCOUT_REPORTER_ENABLED) {
-      // Expose Jest config file path via environment variables
-      process.env.JEST_CONFIG_PATH = configPath;
+  }
+  let initialOptions: Config.InitialOptions | undefined;
+
+  // If a config path was provided via argv, use that; otherwise we already set resolvedConfigPath above
+  if (argv.config) {
+    try {
+      initialOptions = JSON.parse(argv.config);
+    } catch (err) {
+      resolvedConfigPath = argv.config;
     }
   }
 
-  if (process.env.NODE_ENV == null) {
-    process.env.NODE_ENV = 'test';
+  if (!initialOptions && !resolvedConfigPath) {
+    throw new Error(
+      `--config is not set or invalid, and no config path was found for any listed files`
+    );
   }
 
-  if (SCOUT_REPORTER_ENABLED && argv.config) {
+  if (SCOUT_REPORTER_ENABLED && resolvedConfigPath) {
     // Expose Jest config file path via environment variables
-    process.env.JEST_CONFIG_PATH = argv.config;
+    process.env.JEST_CONFIG_PATH = resolvedConfigPath;
   }
 
-  run().then(() => {
+  if (!initialOptions) {
+    const configFileExists = await Fs.promises
+      .stat(resolvedConfigPath!)
+      .then((stat) => stat.isFile())
+      .catch(() => false);
+
+    if (!configFileExists) {
+      throw new Error(`Config at ${resolvedConfigPath} does not exist`);
+    }
+
+    // readInitialOptions returns an object that includes the resolved Jest config at `config`
+    // along with some metadata (e.g. configPath). We only want to pass the actual Jest
+    // config object to --config, augmented with our overrides.
+    initialOptions = (await readInitialOptions(resolvedConfigPath)).config;
+  }
+  const baseConfig = initialOptions;
+
+  const cacheDirectory = join(REPO_ROOT, 'data', 'jest-cache');
+
+  const inlineConfig = {
+    ...baseConfig,
+    id: 'kbn-test-jest',
+    cacheDirectory,
+  };
+
+  const originalArgv = process.argv.concat();
+
+  const tmpConfigPath = Path.join(cacheDirectory, `${process.pid}.jest.config.json`);
+
+  await Fs.promises.mkdir(cacheDirectory, { recursive: true });
+
+  await Fs.promises.writeFile(tmpConfigPath, JSON.stringify(inlineConfig, null, 2), 'utf8');
+
+  const argvWithoutConfig = withoutFlags(process.argv, 'config');
+
+  // Provide the inline JSON config as the last --config argument so it takes precedence
+  const jestArgv = argvWithoutConfig
```

Comment:
if I understand correctly, most of the edits in the file are around expanding the resolved jest config into JSON, adding the shared cache setup, and passing that config in as an inline config? Could you break up this function to multiple functions (can stay within this file) for readability and to better communicate the behaviour?